# 01 Importing libraries

In [32]:
import pandas as pd
import numpy as np
import os

# 02 data

In [33]:
# Create path for dataframes

In [34]:
path = r'C:\Users\Lenad\Documents\Data Analytics Immersion\Achievement 4\Jupyter folder\Instacart basket analysis'

In [35]:
# Import orders_wrangled.csv from Prepared Data folder

In [36]:
df_ords = pd.read_csv(os.path.join(path, '02. Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [37]:
# Import products.csv from Original Data folder

In [38]:
df_prods = pd.read_csv(os.path.join(path, '02. Data', 'Original Data', 'products.csv'), index_col = False)

In [39]:
# Create a dataframe

df_test = pd.DataFrame()

In [40]:
# Create a mixed type column

df_test['mix'] = ['a', 'b', 1, True]

In [41]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [42]:
# Check for mixed types

for col in df_test.columns.tolist():
    weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_test[weird]) > 0:
        print(col)

mix


In [43]:
# Find missing values

df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [44]:
# Looking for null values

df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [45]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [46]:
# Checking number of rows and columns in df_prods dataframe

df_prods.shape

(49693, 5)

In [47]:
# Create new dataframe without missing values

df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [48]:
# Checking number of rows and columns in df_prods_clean dataframe

df_prods_clean.shape

(49677, 5)

In [49]:
# Checking for full duplicates in df_prods_clean dataframe

df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [50]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [51]:
# Checking number of rows and columns in df_prods_clean dataframe

df_prods_clean.shape

(49677, 5)

In [52]:
# Removing duplicates

df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [53]:
df_prods_clean_no_dups

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49672 rows x 5 columns]

# 03 Exporting data

In [54]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02. Data', 'Prepared Data', 'products_checked.csv'))

# 04 Task 4.5

### 2. Run the df.describe function on the df_ords dataframe.

In [55]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_last_order  
count        3.421083e+06       3.421083e+06           3.214874e+06  
mean         2.776219e+00       1.345202e+01           1.111484e+01  
std          2.046829e+00       4.226088e+00           9.206737e+00  
min          0.000000e+00       0.000000e+00           0.000000e+00  
25%          1.000000e+00       1.000000e+01           4.000000e+00  
50%          3.000000e+00       1.300000e+01           7.000000e+00  
75%          5.000000e+00       1.600000e+01           1.500000e+01  
max          6.000000e+00       2.300000e+01           3.000000e+01

Unnamed - not needed, duplicate of order_id column.

user_id - has a count that is lower than the maximum, presumably because several orders were made by the same users.

order_number - max is lower than the count, might be worth investigating further.

orders_day_of_week - difference of 7 between min and max, as expected.

order_hour_of_day - difference of 24 between min and max, as expected.

days_since_last_order - only goes up to 30 days, worth investigating whether that is unusual or a built-in cut-off.

### 3. Check for mixed-type data in the df_ords dataframe

In [56]:
# Checking for mixed data types

for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

No mixed data types were found.

### 5. Run a check for missing values in the df_ords dataframe

In [57]:
# Checking for missing values in df_ords dataframe

df_ords.isnull().sum()

Unnamed: 0                    0
order_id                      0
user_id                       0
eval_set                      0
order_number                  0
orders_day_of_week            0
order_hour_of_day             0
days_since_last_order    206209
dtype: int64

There are 206,209 missing values in the days_since_last_order column.
However, this could be explained by those placing an order for the first time, and so is unlikely to be of concern.

### 6. Address the missing values

In [62]:
# Creating a new variable in which to see the missing values, named df_first_order

df_first_order = df_ords[df_ords['days_since_last_order'].isnull() == True]

In [60]:
df_first_order

Unnamed: 0  order_id  user_id eval_set  order_number  \
0                 0   2539329        1    prior             1   
11               11   2168274        2    prior             1   
26               26   1374495        3    prior             1   
39               39   3343014        4    prior             1   
45               45   2717275        5    prior             1   
...             ...       ...      ...      ...           ...   
3420930     3420930    969311   206205    prior             1   
3420934     3420934   3189322   206206    prior             1   
3421002     3421002   2166133   206207    prior             1   
3421019     3421019   2227043   206208    prior             1   
3421069     3421069   3154581   206209    prior             1   

         orders_day_of_week  order_hour_of_day  days_since_last_order  
0                         2                  8                    NaN  
11                        2                 11                    NaN  
26                        1                 14                    NaN  
39                        6                 11                    NaN  
45                        3                 12                    NaN  
...                     ...                ...                    ...  
3420930                   4                 12                    NaN  
3420934                   3                 18                    NaN  
3421002                   6                 19                    NaN  
3421019                   1                 15                    NaN  
3421069                   3                 11                    NaN  

[206209 rows x 8 columns]

It seems as if all the orders with missing values in the days_since_last_order column have an order number of '1', suggesting that these are indeed first orders. However, it would be worth checking that this is the case for the whole column.

In [61]:
# Checking order_number value of '1' for entire df_first_order column

df_first_order['order_number'].value_counts()

order_number
1    206209
Name: count, dtype: int64

Checking the frequency of the value of '1' in the order_number column confirms that only one order has been made by these users, meaning there would have been 0 days since their last order.

This has addressed the missing values in the dataframe to date, but in order to filter relevant data into this column in the future, specific code would have to be written.

### 7. Run a check for duplicate values in df_ords

In [64]:
# Checking for duplicates in df_ords

df_dups = df_ords[df_ords.duplicated()]

In [65]:
df_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, orders_day_of_week, order_hour_of_day, days_since_last_order]
Index: []

In [66]:
df_dups.shape

(0, 8)

There are no duplicates in the df_ords dataframe.

### 8. Address the duplicates

If I had found full duplicates and the client had confirmed there was no requirement for the duplicates, I would: 

Check the number of rows in the df_ords dataframe using df_ords.shape

Create a new dataframe that does not include the duplicates - df_ords_no_dups = df_ords.drop_duplicates()

Check that the new dataframe has the same number fewer rows than duplicates found - df_ords_no_dups.shape

# Export data

### 9. Export cleaned df_prods and df_ords data as .csv files

In [67]:
# Export cleaned df_ords data as 'orders_wrangled.csv'

df_ords.to_csv(os.path.join(path, '02. Data','Prepared Data', 'orders_wrangled.csv'))

In [68]:
# Export cleaned df_prods data as 'products_checked.csv'

df_prods.to_csv(os.path.join(path, '02. Data','Prepared Data', 'products_checked.csv'))